test/csuite/random_directio/main.c

```diff
@@ -0,0 +1,1127 @@
+thread_run(void *arg)
+{
+	WT_CURSOR *cursor, *rev, *sch;
+	WT_RAND_STATE rnd;
+	WT_SESSION *session;
+	WT_THREAD_DATA *td;
+	size_t lsize;
+	uint64_t i;
+	uint32_t kvsize;
+	int ret;
+	char *buf1, *buf2;
+	char large[LARGE_WRITE_SIZE];
+	bool retryable_error;
+
+	__wt_random_init(&rnd);
+	lsize = sizeof(large);
+	memset(large, 0, lsize);
+
+	td = (WT_THREAD_DATA *)arg;
+	large_buf(large, lsize, td->id, true);
+
+	testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+	testutil_check(session->open_cursor(session, uri_main, NULL, NULL,
+	    &cursor));
+	testutil_check(session->open_cursor(session, uri_rev, NULL, NULL,
+	    &rev));
+
+	/*
+	 * Split the allocated buffer into two parts, one for
+	 * the key, one for the value.
+	 */
+	kvsize = td->datasize / 2;
+	buf1 = td->data;
+	buf2 = &td->data[kvsize];
+
+	/*
+	 * Continuing writing until we're killed.
+	 */
+	printf("Thread %" PRIu32 "\n", td->id);
+	for (i = 0; ; ++i) {
+	again:
+		retryable_error = false;
+		/*
+		if (i > 0 && i % 10000 == 0)
+			printf("Thread %d completed %d entries\n",
+			    (int)td->id, (int)i);
+		*/
+
+		gen_kv(buf1, kvsize, i, td->id, large, true);
+		gen_kv(buf2, kvsize, i, td->id, large, false);
+
+		testutil_check(session->begin_transaction(session, NULL));
+		cursor->set_key(cursor, buf1);
+		/*
+		 * Every 1000th record write a very large value that exceeds the
+		 * log buffer size.  This forces us to use the unbuffered path.
+		 */
+		if (i % 1000 == 0) {
+			cursor->set_value(cursor, large);
+		} else {
+			cursor->set_value(cursor, buf2);
+		}
+		testutil_check(cursor->insert(cursor));
+
+		/*
+		 * The reverse table has no very large records.
+		 */
+		rev->set_key(rev, buf2);
+		rev->set_value(rev, buf1);
+		testutil_check(rev->insert(rev));
+
+		/*
+		 * If we are doing a schema test, generate operations
+		 * for additional tables.  Each table has a 'lifetime'
+		 * of 4 values of the id.
+		 */
+		if (td->schema_test) {
+			if (!retryable_error && SCHEMA_OP(i, 0)) {
+				/* Create a table. */
+				gen_table_name(buf1, kvsize, i, td->id);
+				testutil_check(session->create(session, buf1,
+				    "key_format=S,value_format=S"));
+			}
+			if (!retryable_error && SCHEMA_OP(i, 1)) {
+				/* Insert a value into the table. */
+				gen_table_name(buf1, kvsize, i - 1, td->id);
+				testutil_check(session->open_cursor(session,
+				    buf1, NULL, NULL, &sch));
+				sch->set_key(sch, buf1);
+				sch->set_value(sch, buf1);
+				testutil_check(sch->insert(sch));
+				sch->close(sch);
+			}
+			if (!retryable_error && SCHEMA_OP(i, 2)) {
+				/* Rename the table. */
+				gen_table_name(buf1, kvsize, i - 2, td->id);
+				gen_table2_name(buf2, kvsize, i - 2, td->id);
+				/*
+				 * XXX
+				 * We notice occasional EBUSY errors from
+				 * rename, even though neither URI should be
+				 * used by any other thread.
+				 */
+				/*
+				printf(" rename(\"%s\", \"%s\")\n", buf1, buf2);
+				*/
+				if ((ret = session->rename(session, buf1, buf2,
```

Comment:
Just FYI, until WT-2329 is fixed there is a definitely possibility of files/metadata out of sync with `rename`. A crash at the wrong time of rename can leave things orphaned.

Reply:
So is the plan to change this test to avoid exercising rename until WT-2329 is fixed?

Reply:
I've modified the test to optionally do rename, and actually to control more what gets done and what gets checked. I find that we can do both create and drop and reliably finish recovery in all situations (methods set to {dsync,fsync,none}) but we can't expect to see the created or dropped tables to appear or disappear at the right moment.  So those options with be in smoke.sh, with the various things we can't yet do commented out.